File: Ejercicios Propuestos 1/Ejercicio 24/EJERCICIO_24.cpp
#include <iostream>
using namespace std;
int main() {
    int num_postulantes, i;
    int corr_rmv, inc_rmv, corr_cie, inc_cie, corr_let, inc_let;
    int cant_ingresantes;
    double nota_final, suma_notas, suma_notas_cuadrado;
    double media, varianza, nota_min_ing, nota_max_ing;
    bool tiene_ingresantes;
      num_postulantes = 5; 
    suma_notas = 0;
    suma_notas_cuadrado = 0;
    cant_ingresantes = 0;
    tiene_ingresantes = false;
    nota_min_ing = 999;
    nota_max_ing = -999;

    for (i = 1; i <= num_postulantes; i++) {
        cout << "==================================================" << endl;
        cout << "PROCESANDO POSTULANTE N* " << i << endl;
        cout << "==================================================" << endl;
        
        cout << "Parte 1: Razonamiento Mat. y Verbal (Max 40)" << endl;
        cout << " - Cantidad de respuestas CORRECTAS: ";
        cin >> corr_rmv;
        cout << " - Cantidad de respuestas INCORRECTAS: ";
        cin >> inc_rmv;
        
        cout << "Parte 2: Conocimiento en Ciencias (Max 30)" << endl;
        cout << " - Cantidad de respuestas CORRECTAS: ";
        cin >> corr_cie;
        cout << " - Cantidad de respuestas INCORRECTAS: ";
        cin >> inc_cie;
        
        cout << "Parte 3: Conocimiento en Letras (Max 30)" << endl;
        cout << " - Cantidad de respuestas CORRECTAS: ";
        cin >> corr_let;
        cout << " - Cantidad de respuestas INCORRECTAS: ";
        cin >> inc_let;

        nota_final = (corr_rmv * 2.0 - inc_rmv * 1.0) + 
                     (corr_cie * 3.0 - inc_cie * 1.5) + 
                     (corr_let * 1.0 - inc_let * 0.5);

        suma_notas += nota_final;
        suma_notas_cuadrado += (nota_final * nota_final);

        cout << " -> Nota final del postulante: " << nota_final << endl;

        if (nota_final >= 120.0 && cant_ingresantes < 150) {
            cout << " -> RESULTADO: !INGRESO!" << endl;
            cant_ingresantes++;
            tiene_ingresantes = true;

            if (nota_final < nota_min_ing) {
                nota_min_ing = nota_final;
            }
            if (nota_final > nota_max_ing) {
                nota_max_ing = nota_final;
            }
        } else {
            cout << " -> RESULTADO: NO INGRESO" << endl;
        }
        cout << endl;
    }

    media = suma_notas / num_postulantes;
    varianza = (suma_notas_cuadrado / num_postulantes) - (media * media);

    cout << "==================================================" << endl;
    cout << "                REPORTE GENERAL                   " << endl;
    cout << "==================================================" << endl;
    cout << "b) Media aritmetica de las notas: " << media << endl;
    cout << "c) Varianza de las notas: " << varianza << endl;
    
    if (tiene_ingresantes) {
        cout << "d) Reporte de Ingresantes: " << endl;
        cout << "   - Vacantes ocupadas: " << cant_ingresantes << endl;
        cout << "   - Nota MINIMA entre ingresantes: " << nota_min_ing << endl;
        cout << "   - Nota MAXIMA entre ingresantes: " << nota_max_ing << endl;
    } else {
        cout << "d) Reporte de Ingresantes: Ningun postulante logro ingresar." << endl;
    }
    cout << "==================================================" << endl;

    return 0;
}

File: Ejercicios Propuestos 1/Ejercicio 24/Ejercicio 24.jl
# Función auxiliar para leer y validar que la entrada sea un número entero
function leer_entero(mensaje::String)::Int
    while true
        print(mensaje)
        entrada = readline()
        # Intentamos convertir la entrada a entero
        numero = tryparse(Int, entrada)
        if numero !== nothing
            return numero
        else
            println("Error: Por favor, ingrese un número entero válido.")
        end
    end
end

function main()
    println("Ingrese la cantidad total de postulantes a evaluar: ")
    num_postulantes = leer_entero("")
    
    while num_postulantes <= 0
        println("Error: La cantidad debe ser mayor a 0. Intente de nuevo: ")
        num_postulantes = leer_entero("")
    end
    # --------------------------------------------
    
    suma_notas = 0.0
    suma_notas_cuadrado = 0.0
    cant_ingresantes = 0
    tiene_ingresantes = false
    nota_min_ing = 999.0  
    nota_max_ing = -999.0 
    
    for i in 1:num_postulantes
        println("==================================================")
        println("PROCESANDO POSTULANTE Nº ", i)
        println("==================================================")
        
        println("Parte 1: Razonamiento Mat. y Verbal (Máx 40)")
        corr_rmv = leer_entero("  - Cantidad de respuestas CORRECTAS: ")
        inc_rmv = leer_entero("  - Cantidad de respuestas INCORRECTAS: ")
        
        println("Parte 2: Conocimiento en Ciencias (Máx 30)")
        corr_cie = leer_entero("  - Cantidad de respuestas CORRECTAS: ")
        inc_cie = leer_entero("  - Cantidad de respuestas INCORRECTAS: ")
        
        println("Parte 3: Conocimiento en Letras (Máx 30)")
        corr_let = leer_entero("  - Cantidad de respuestas CORRECTAS: ")
        inc_let = leer_entero("  - Cantidad de respuestas INCORRECTAS: ")
        
        # Cálculo de la nota final
        nota_final = (corr_rmv * 2) - (inc_rmv * 1) + (corr_cie * 3) - (inc_cie * 1.5) + (corr_let * 1) - (inc_let * 0.5)
        
        println("--------------------------------------------------")
        println("-> NOTA FINAL: ", nota_final)
        
        # Evaluación del ingreso (Mínimo 120 puntos y límite de 150 vacantes)
        if nota_final >= 120 && cant_ingresantes < 150
            println("-> ESTATUS: ¡INGRESÓ EL POSTULANTE!")
            cant_ingresantes += 1
            
            if !tiene_ingresantes
                nota_min_ing = nota_final
                nota_max_ing = nota_final
                tiene_ingresantes = true
            else
                if nota_final < nota_min_ing
                    nota_min_ing = nota_final
                end
                if nota_final > nota_max_ing
                    nota_max_ing = nota_final
                end
            end
        else
            if nota_final >= 120 && cant_ingresantes >= 150
                println("-> ESTATUS: NO INGRESÓ (Alcanzó puntaje pero no hay vacantes disponibles)")
            else 
                println("-> ESTATUS: NO INGRESÓ (Puntaje menor al mínimo de 120)")
            end
        end
        
        suma_notas += nota_final
        suma_notas_cuadrado += (nota_final * nota_final)
        println()
    end
    
    println("==================================================")
    println("            REPORTE GENERAL DE ADMISIÓN           ")
    println("==================================================")
    
    media = suma_notas / num_postulantes
    println("b) Media aritmética de todos los postulantes: ", media)
    
    varianza = (suma_notas_cuadrado / num_postulantes) - (media * media)
    println("c) Varianza de las notas de los postulantes: ", varianza)
    
    if tiene_ingresantes
        println("d) Reporte de Ingresantes:")
        println("   - Total de vacantes ocupadas: ", cant_ingresantes)
        println("   - Nota MÍNIMA entre ingresantes: ", nota_min_ing)
        println("   - Nota MÁXIMA entre ingresantes: ", nota_max_ing)
    else
        println("d) Reporte de Ingresantes: Ningún postulante logró ingresar.")
    end
    println("==================================================")
end

# Ejecutar el algoritmo principal
main()
